src/frontend/src/binder/set_expr.rs

```diff
@@ -65,16 +99,36 @@ impl Binder {
         match set_expr {
             SetExpr::Select(s) => Ok(BoundSetExpr::Select(Box::new(self.bind_select(*s)?))),
             SetExpr::Values(v) => Ok(BoundSetExpr::Values(Box::new(self.bind_values(v, None)?))),
-            SetExpr::Query(q) => Err(ErrorCode::NotImplemented(
-                format!("Parenthesized SELECT subquery: ({:})\nYou can try to remove the parentheses if they are optional", q),
-                3584.into(),
-            )
-            .into()),
-            SetExpr::SetOperation { .. } => Err(ErrorCode::NotImplemented(
-                format!("set expr: {:}", set_expr),
-                None.into(),
-            )
-            .into()),
+            SetExpr::Query(q) => Ok(BoundSetExpr::Query(Box::new(self.bind_query(*q)?))),
+            SetExpr::SetOperation {
+                op,
+                all,
+                left,
+                right,
+            } => {
+                match op {
+                    SetOperator::Union => {
+                        let left = Box::new(self.bind_set_expr(*left)?);
+                        let mut new_context = BindContext::default();
+                        // Swap context for the right side.
+                        std::mem::swap(&mut self.context, &mut new_context);
+                        let right = Box::new(self.bind_set_expr(*right)?);
+                        // Swap context back to the left side.
+                        std::mem::swap(&mut self.context, &mut new_context);
+                        Ok(BoundSetExpr::SetOperation {
+                            op: BoundSetOperation::UNION,
+                            all: all,
+                            left: left,
+                            right: right,
+                        })
```

Comment:
The data types of ~~context~~/schema should be the least_restrictive of both sides. We may require them to be the same types now and add casting support in a followup.

As for context, an earlier design (https://github.com/risingwavelabs/risingwave/pull/2329#discussion_r866476756) is to leave it as empty, so that `order by` can reject expressions referencing input columns:
```
test=# select 1 as a union all select 2 as b order by a+1;
ERROR:  invalid UNION/INTERSECT/EXCEPT ORDER BY clause
LINE 1: select 1 as a union all select 2 as b order by a+1;
                                                       ^
DETAIL:  Only result column names can be used, not expressions or functions.
HINT:  Add the expression/function to every SELECT, or move the UNION into a FROM clause.
```

Reply:
I think currently, our behavior is the same as PG except the error message.
```sql
dev=> select 1 as a union all select 2 as b order by a+1;
ERROR:  QueryError: Item not found: Invalid column: a
dev=> select 1 as a union all select 2 as b order by a;
 a
---
 1
 2
(2 rows)
```

Reply:
My bad. The counterexample is not strong enough. Try this instead:
```
create table t1 (a int);
create table t2 (b int);
select a from t1 union select b from t2 order by a+1;
```

Reply:
> The data types of ~context~/schema should be the least_restrictive of both sides. We may require them to be the same types now and add casting support in a followup.
> 
> As for context, an earlier design ([#2329 (comment)](https://github.com/risingwavelabs/risingwave/pull/2329#discussion_r866476756)) is to leave it as empty, so that `order by` can reject expressions referencing input columns:
> 
> ```
> test=# select 1 as a union all select 2 as b order by a+1;
> ERROR:  invalid UNION/INTERSECT/EXCEPT ORDER BY clause
> LINE 1: select 1 as a union all select 2 as b order by a+1;
>                                                        ^
> DETAIL:  Only result column names can be used, not expressions or functions.
> HINT:  Add the expression/function to every SELECT, or move the UNION into a FROM clause.
> ```

IIUC, we can reset the context for right side and set operation itself. I also have some test cases in e2e test.

Reply:
Yes. What about asserting the context before binding left is also empty?